utils/fork-tree/src/lib.rs

```diff
@@ -226,29 +142,35 @@ where
 			}
 		}
 
-		for root in self.roots.iter_mut() {
-			if root.hash == hash {
+		let mut children = &mut self.roots;
+		let mut len = children.len();
+		let mut i = 0;
+		while i < len {
+			let child = &children[i];
+			if child.hash == hash {
 				return Err(Error::Duplicate)
 			}
-
-			match root.import(hash, number, data, is_descendent_of)? {
-				Some((h, n, d)) => {
-					hash = h;
-					number = n;
-					data = d;
-				},
-				None => {
-					self.rebalance();
-					return Ok(false)
-				},
+			if child.number < number && is_descendent_of(&child.hash, &hash)? {
+				children = &mut children[i].children;
+				len = children.len();
+				i = 0;
+			} else {
+				i += 1;
 			}
 		}
 
-		self.roots.push(Node { data, hash, number, children: Vec::new() });
+		let is_first = children.is_empty();
+		children.push(Node { data, hash, number, children: Default::default() });
 
-		self.rebalance();
+		// Quick way to check if the pushed node is a root
+		let is_root = children.as_ptr() == self.roots.as_ptr();
 
-		Ok(true)
+		if is_first {
+			// Rebalance is required only if we've extended the branch depth.
+			self.rebalance();
+		}
+
+		Ok(is_root)
 	}
```

Comment:
If is not a problem I really prefer to keep it within the `ForkTree impl`  to keep it consistent with the rest of the methods.

The only method that is within the `Node` impl is the max_depth because it refers to the node itself.